aptos-move/mvhashmap/src/code_cache/sync.rs

```diff
@@ -0,0 +1,249 @@
+            let lock = shard.write();
+            locked_cache_shards.push(lock);
+        }
+
+        // At this point all shards are locked. Only one thread can manipulate the locked cache.
+        LockedSyncModuleCache {
+            cache: &self.cache,
+            phantom_data: self.phantom_data,
+            locked_cache_shards: RefCell::new(locked_cache_shards),
+        }
+    }
+
+    pub fn filter_into<T, P, F>(&self, collector: &mut HashMap<K, T>, p: P, f: F)
+    where
+        P: Fn(&M) -> bool,
+        F: Fn(&M) -> T,
+    {
+        for r in self.cache.iter().filter(|r| p(r.value())) {
+            collector.insert(r.key().clone(), f(r.value()));
+        }
+    }
+}
+
+impl<K, M, E> ModuleCache for SyncModuleCache<K, M, E>
+where
+    K: Eq + Hash + Clone,
+    M: Clone,
+{
+    type Error = E;
+    type Key = K;
+    type Module = M;
+
+    fn store_module(&self, key: Self::Key, module: Self::Module) {
+        self.cache.insert(key, CachePadded::new(module));
+    }
+
+    fn fetch_module_or_store_with<F>(
+        &self,
+        key: &Self::Key,
+        default: F,
+    ) -> Result<Option<Self::Module>, Self::Error>
+    where
+        F: FnOnce() -> Result<Option<Self::Module>, Self::Error>,
+    {
+        if let Some(m) = self.cache.get(key) {
+            return Ok(Some(m.value().deref().clone()));
+        }
+
+        // This takes the write lock!
+        match self.cache.entry(key.clone()) {
+            Entry::Occupied(entry) => {
+                // In case the entry is already initialized, return it. We must not re-initialize
+                // it with the base value because there can be a transaction committing this value
+                // at the same time and caching it first. Hence, that "newer" value needs to stay.
+                Ok(Some(entry.get().deref().clone()))
+            },
+            Entry::Vacant(entry) => match default()? {
+                Some(m) => {
+                    entry.insert(CachePadded::new(m.clone()));
+                    Ok(Some(m))
+                },
+                None => Ok(None),
+            },
+        }
+    }
+}
+
+pub type HashMapShard<K, M> = HashMap<K, dashmap::SharedValue<CachePadded<M>>, RandomState>;
+pub type HashMapShardWriteGuard<'a, K, M> =
+    lock_api::RwLockWriteGuard<'a, dashmap::RawRwLock, HashMapShard<K, M>>;
+
+pub struct LockedSyncModuleCache<'a, K, M, E> {
+    // Note: the reference to the dashmap is used ONLY to calculate the shard index!
+    cache: &'a DashMap<K, CachePadded<M>>,
+    phantom_data: PhantomData<E>,
+    locked_cache_shards: RefCell<Vec<HashMapShardWriteGuard<'a, K, M>>>,
+}
+
+impl<'a, K, M, E> LockedSyncModuleCache<'a, K, M, E>
+where
+    K: Eq + Hash + Clone,
+    M: Clone,
+{
+    /// Unlocks the module cache.
+    pub fn unlock(self) {
+        for lock in self.locked_cache_shards.into_inner() {
+            drop(lock)
+        }
+    }
+}
+
+impl<K, M, E> ModuleCache for LockedSyncModuleCache<'_, K, M, E>
```

Comment:
As discussed, let's get rid of LockedSyncModuleCache and instead have global verification lock, as well as disallow overwriting verified earlier version to a newly published higher committed version - in that case some error or return value could indicate to the caller to retry verification. Should hopefully simplify things